clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+      SatisfiabilityResult rhsResult = computeSatisfiability(rhsVar, Env2);
+
+      if (lhsResult == SR::Top || rhsResult == SR::Top) 
+        return MergedEnv.makeTopBoolValue();
+      
+      if (lhsResult == SR::Nullptr && rhsResult == SR::Nullptr)
+        return MergedEnv.makeAtomicBoolValue();
+
+      if (lhsResult == SR::True || rhsResult == SR::True)
+        return MergedEnv.getBoolLiteralValue(true);
+      else if (lhsResult == SR::False && rhsResult == SR::False)
+        return MergedEnv.getBoolLiteralValue(false);
+
+      return MergedEnv.makeTopBoolValue();
+    };
+
+    // Attempt 2: Slow evaluation, but only produces simple variables.
+    const auto MergeValues = [&](llvm::StringRef Name) -> BoolValue & {
+      BoolValue *lhsVar = cast_or_null<BoolValue>(Val1.getProperty(Name));
+      BoolValue *rhsVar = cast_or_null<BoolValue>(Val2.getProperty(Name));
+
+      SatisfiabilityResult lhsResult = computeSatisfiability(lhsVar, Env1);
+      SatisfiabilityResult rhsResult = computeSatisfiability(rhsVar, Env2);
+
+      // Handle special cases.
+      if (lhsResult == SR::Top || rhsResult == SR::Top) {
+        return MergedEnv.makeTopBoolValue();
+      } else if (lhsResult == rhsResult) {
+        switch (lhsResult) {
+          case SR::Nullptr:
+            return MergedEnv.makeAtomicBoolValue();
+          case SR::Top:
+            return *lhsVar;
+          case SR::True:
+            return MergedEnv.getBoolLiteralValue(true);
+          case SR::False:
+            return MergedEnv.getBoolLiteralValue(false);
+          case SR::Unknown:
+            if (MergedEnv.flowConditionImplies(MergedEnv.arena().makeEquals(lhsVar->formula(), rhsVar->formula())))
+              return *lhsVar;
+            
+            return MergedEnv.makeTopBoolValue();
+        }
+      }
+
+      return MergedEnv.makeTopBoolValue();
+    };
+
+    // Attempt 3: Merge to (L or R), a more complex variable
+    const auto SlowMergeValues = [&](llvm::StringRef Name) -> BoolValue & {
+      BoolValue *lhsVar = cast_or_null<BoolValue>(Val1.getProperty(Name));
+      BoolValue *rhsVar = cast_or_null<BoolValue>(Val2.getProperty(Name));
+
+      SatisfiabilityResult lhsResult = computeSatisfiability(lhsVar, Env1);
+      SatisfiabilityResult rhsResult = computeSatisfiability(rhsVar, Env2);
+
+      // Handle special cases.
+      if (lhsResult == SR::Top || rhsResult == SR::Top) {
+        return MergedEnv.makeTopBoolValue();
+      } else if (lhsResult == rhsResult) {
+        switch (lhsResult) {
+          case SR::Nullptr:
+            return MergedEnv.makeAtomicBoolValue();
+          case SR::Top:
+            return *lhsVar;
+          case SR::True:
+            return MergedEnv.getBoolLiteralValue(true);
+          case SR::False:
+            return MergedEnv.getBoolLiteralValue(false);
+          case SR::Unknown:
+            break;
+          default:
+            llvm_unreachable("all cases covered in switch");
```

Comment:
✏️ (My problem with having a `default:` here is that this kills `-Wswitch`, but I guess because not all branches of the switch actually return we can't really do any better than this... ☹️)